File: montelier/dinero.py
"""Cálculo de IVA y formato de importes.

Todos los importes de la base van SIN IVA. El IVA se calcula siempre aquí,
con Decimal y redondeo comercial (mitad hacia arriba) a céntimos.
"""
from decimal import Decimal, ROUND_HALF_UP

TIPO_IVA = Decimal("0.21")
CENTIMO = Decimal("0.01")


def a_decimal(valor) -> Decimal:
    if valor is None:
        return Decimal("0.00")
    if isinstance(valor, Decimal):
        return valor
    # str() evita arrastrar el error binario del float (0.1 -> 0.1000000000000000055...)
    return Decimal(str(valor))


def redondear(valor) -> Decimal:
    return a_decimal(valor).quantize(CENTIMO, rounding=ROUND_HALF_UP)


def desglose(importe_transporte, importe_montaje=None) -> dict:
    """Base, IVA y total de una factura a partir de sus importes sin IVA."""
    transporte = redondear(importe_transporte)
    montaje = redondear(importe_montaje)
    base = transporte + montaje
    iva = redondear(base * TIPO_IVA)
    return {
        "transporte": transporte,
        "montaje": montaje,
        "base": base,
        "iva": iva,
        "total": base + iva,
    }


def sumar_desgloses(desgloses) -> dict:
    """Suma de varias facturas: el IVA total es la suma de los IVA de cada factura
    (ya redondeados), igual que en un libro de facturas emitidas."""
    total = {"transporte": Decimal("0.00"), "montaje": Decimal("0.00"),
             "base": Decimal("0.00"), "iva": Decimal("0.00"), "total": Decimal("0.00")}
    for d in desgloses:
        for clave in total:
            total[clave] += d[clave]
    return total


def formato_eur(valor) -> str:
    """1210 -> '1.210,00 €'"""
    cantidad = redondear(valor)
    signo = "-" if cantidad < 0 else ""
    entero, decimales = f"{abs(cantidad):.2f}".split(".")
    grupos = []
    while len(entero) > 3:
        grupos.insert(0, entero[-3:])
        entero = entero[:-3]
    grupos.insert(0, entero)
    return f"{signo}{'.'.join(grupos)},{decimales} €"


def desglose_formateado(d: dict) -> dict:
    return {clave: formato_eur(valor) for clave, valor in d.items()}

File: netlify/lib/dinero.mjs
// IVA y formato de importes (misma lógica que montelier/dinero.py).
// Se trabaja en céntimos enteros para evitar errores de coma flotante.
// Todos los importes de la base van SIN IVA; el IVA (21 %) se calcula aquí y nunca se guarda.

export const aCentimos = (valor) => (valor == null ? 0 : Math.round(Number(String(valor)) * 100));

// 21 % con redondeo comercial (mitad hacia arriba) a céntimos
export const ivaCentimos = (base) => Math.sign(base) * Math.floor((Math.abs(base) * 21 + 50) / 100);

export function desglose(importeTransporte, importeMontaje = null) {
  const transporte = aCentimos(importeTransporte);
  const montaje = aCentimos(importeMontaje);
  const base = transporte + montaje;
  const iva = ivaCentimos(base);
  return { transporte, montaje, base, iva, total: base + iva };
}

export function sumarDesgloses(lista) {
  const total = { transporte: 0, montaje: 0, base: 0, iva: 0, total: 0 };
  for (const d of lista) for (const k of Object.keys(total)) total[k] += d[k];
  return total;
}

// 121000 céntimos -> "1.210,00 €"
export function formatoEur(centimos) {
  const signo = centimos < 0 ? "-" : "";
  const abs = Math.abs(Math.round(centimos));
  const entero = String(Math.floor(abs / 100)).replace(/\B(?=(\d{3})+(?!\d))/g, ".");
  return `${signo}${entero},${String(abs % 100).padStart(2, "0")} €`;
}

export const desgloseFormateado = (d) =>
  Object.fromEntries(Object.entries(d).map(([k, v]) => [k, formatoEur(v)]));
